src/plugin.rs

```diff
@@ -0,0 +1,218 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use pyo3::{exceptions::PyAttributeError, prelude::*, types::PyType};
+
+use crate::{DecryptError, EncryptError, IdentityError, RecipientError};
+
+/// Hack, because the orphan rule would prevent us from deriving a
+/// foreign trait on a foreign object. Instead, define a newtype.
+///
+/// Inner type is PyAny, because we do duck-typing at runtime, and
+/// declaring a protocol in the type stubs.
+#[derive(Clone)]
+pub(crate) struct PyCallbacks(Py<PyAny>);
+
+impl PyCallbacks {
+    fn new(inner: Py<PyAny>, py: Python<'_>) -> PyResult<Self> {
+        {
+            let inner = inner.as_ref(py);
+
+            for method_name in [
+                "display_message",
+                "confirm",
+                "request_public_string",
+                "request_passphrase",
+            ] {
+                // I would use `hasattr`, but it doesn't throw, and
+                // constructing pretty AttributeErrors is incredibly
+                // hard.
+                //
+                // XXX: Should we even throw it? Better ask for
+                // forgiveness than permission, and we're panicking
+                // when things get called. And static typecheckers
+                // will catch a protocol violation.
+                inner.getattr(method_name)?;
+            }
+        }
+        Ok(Self(inner))
+    }
+}
+
+// Since we have no way to pass errors from these callbacks, we might
+// as well panic.
+//
+// These callbacks don't look like they're supposed to fail anyway.
+impl age::Callbacks for PyCallbacks {
+    fn display_message(&self, message: &str) {
+        Python::with_gil(|py| {
+            self.0
+                .call_method1(py, pyo3::intern!(py, "display_message"), (message,))
+                .expect("`display_message` callback error")
```

Comment:
This looks acceptable to me, given that the `age::Callbacks` trait doesn't seem to allow errors.

To exhaust one last option: have you tried using [`age_plugin`](https://docs.rs/age-plugin/0.4.0/age_plugin/index.html)? It seems to define its own (fallible) `Callbacks` trait, which may solve our problem here.